### DISCUSSION TITLE
convert -fx: different behavior with v7 vs v6

### DISCUSSION THREAD

**Discussion by herrriehm** [ImageMagick/ImageMagick#5892] [Help] [answered]
I have this infile.png:

```
identify infile.png
infile.png PNG 320x240 320x240+0+0 16-bit Grayscale Gray 69372B 0.000u 0:00.000
```
![infile](https://user-images.githubusercontent.com/1580028/208509186-92044c6c-19a6-48f4-a541-acd1d3254598.png)

I want to run the following command:

```
convert infile.png -fx "(1411.5/ln(15670.33/(0.011282093*((65535*u-4717)?(65535*u-4717):1))+1)-275.40539015946)/63.360868248809" outfile.png
```

IM 6.9.12-65 Q16-HDRI x64 83634ca:20221008 gives me the right image:
![outfile_V6](https://user-images.githubusercontent.com/1580028/208509122-67da05f9-7297-49e6-b23f-373670a6894b.png)

saved as txt file:

```
# ImageMagick pixel enumeration: 320,240,65535,gray
0,0: (576.769,576.769,576.769)  #024102410241  gray(0.880094%)
1,0: (656.034,656.034,656.034)  #029002900290  gray(1.00104%)
2,0: (748.42,748.42,748.42)  #02EC02EC02EC  gray(1.14202%)
3,0: (840.712,840.712,840.712)  #034903490349  gray(1.28284%)
4,0: (906.577,906.577,906.577)  #038B038B038B  gray(1.38335%)
5,0: (913.161,913.161,913.161)  #039103910391  gray(1.39339%)
...
```

IM 7.1.0-55 Q16-HDRI x64 6edfae3:20221217 gives me a wrong image:
![outfile_V7](https://user-images.githubusercontent.com/1580028/208509146-9ded3982-4e2a-4d2b-95b3-8c4ef79b77da.png)

saved as txt file:

```
# ImageMagick pixel enumeration: 320,240,0,65535,gray
0,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
1,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
2,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
3,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
4,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
5,0: (1459933463506321408)  #FFFFFFFFFFFF  gray(2.22772e+15%)
...
```

I suspect the fx function to be the bad guy here. Something about the behavior of the `u` symbol. I have already read the [Porting Guide](https://imagemagick.org/script/porting.php), but I was not able to find a solution.

May any of you nice people help me? Thank you very much in advance!

**Comment by fmw42** [comment, marked as answer]
Works fine for me on IM 6.9.12-70 and IM 7.1.0-55 on Mac OSX Monterey. Both results are the same as your good one. Note for IM 7, use magick not convert.  What is your OS/platform?

```
convert infile.png -fx "(1411.5/ln(15670.33/(0.011282093*((65535*u-4717)?(65535*u-4717):1))+1)-275.40539015946)/63.360868248809" outfile6.png

magick infile.png -fx "(1411.5/ln(15670.33/(0.011282093*((65535*u-4717)?(65535*u-4717):1))+1)-275.40539015946)/63.360868248809" outfile7.png
```

![outfile6](https://user-images.githubusercontent.com/13458218/208517800-96f1dc8a-258b-4bf4-b6d6-e2d8c1afef53.png)
![outfile7](https://user-images.githubusercontent.com/13458218/208517803-0622776c-9ddd-4f2c-8e99-808e3820c257.png)

**Comment by snibgo** [comment]
For IM v7, I suggest using "magick", not "convert", because "convert" will use the old v6 syntax.

I can't reproduce the problem with IM v7.1.0-54 or 7.1.0-49 (with "magick"). The image is as expected, and the first few "`txt:`" values are:

```
# ImageMagick pixel enumeration: 320,240,0,65535,gray
0,0: (577)  #024102410241  gray(0.880094%)
1,0: (656)  #029002900290  gray(1.00104%)
2,0: (748)  #02EC02EC02EC  gray(1.14202%)
3,0: (841)  #034903490349  gray(1.28284%)
4,0: (907)  #038B038B038B  gray(1.38335%)
5,0: (913)  #039103910391  gray(1.39339%)
```

Your output contains colour values of "`gray(2.22772e+15%)`" which is really weird.

**Comment by snibgo** [comment]
The result of "`gray(2.22772e+15%)`" can be obtained when we are taking a "`ln`" of a negative number. This can happen for your expression when the input pixel value "`u`" is less than 0.0706...

**Comment by herrriehm** [comment]
@fmw42 and @snibgo , please excuse my delayed answer. Thank you very much for your suggestions!

You were right: When I use `magick` without any `convert`, I get the same results in v6 and v7. 🙂

But this success leads me directly to the next difference, which i somehow can't get resolved. The following command: 

```
convert.exe "outfile.png" "palette.png" -clut "embedded1.png" +swap -compose overlay -composite "outfile2.png"
```

works perfect with v6 `convert`:

![outfile2_v6convert](https://user-images.githubusercontent.com/1580028/210274098-1118bfa0-7f62-4776-bcee-bc4e3cc45651.png)

But when I use v7 `magick` instead:

![outfile2_v7magick](https://user-images.githubusercontent.com/1580028/210274110-73dc3aca-0ba6-460a-a011-fffb00485921.png)

Can you nudge me in the right direction once more?

---

Here are the files you need for the command above: 

- `outfile.png`:
![outfile](https://user-images.githubusercontent.com/1580028/210273438-9c734e94-67c3-4cdd-9a11-6fb5aa333e48.png)

- `palette.png`:
![palette](https://user-images.githubusercontent.com/1580028/210273362-3883beb4-1e39-4813-b2e4-d52aeeddb8ce.png)

- `embedded1.png`:
![embedded1](https://user-images.githubusercontent.com/1580028/210274123-a09bda6b-bb31-46dc-964f-94fa6c482b3c.png)

--- 

Versions are:

- ImageMagick 6.9.12-65 Q16-HDRI x64 83634ca:20221008
- ImageMagick 7.1.0-55 Q16-HDRI x64 6edfae3:20221217
- on Windows 10 Pro

**Reply by fmw42** [reply]
add -colorspace sRGB just before the output

**Comment by herrriehm** [comment]
thank you once again @fmw42 ! 

I got it working perfectly on IM7 with the following command:

```
magick.exe "outfile.png" "palette.png" -clut "embedded1.png" +swap -compose overlay -colorspace sRGB -composite "outfile2.png"
```
